### PR TITLE
add position_ids

### PR DESCRIPTION
**Pull request by jiqing-feng**
Some models like gpt2 and llama have "position_ids" in their generation inputs. We can add "position_ids" in the model's config to fix it.

@echarlaix @fxmarty Would you please help to review it? Thanks!

### CODE REVIEW

**Comment by fxmarty** [comment]
Hi @jiqing-feng , just to be sure, is this concerning the ONNX export or ONNX Runtime integration? Or both? In the ONNX export, we should give the option, yes.

**Comment by echarlaix** [comment]
Currently `position_ids` is ignored during ORT inference for causal LM https://github.com/huggingface/optimum/blob/v1.8.2/optimum/onnxruntime/modeling_decoder.py#L683, so when enabling it as inputs for the ONNX export we should enable it for `ORTModelForCausalLM` as well

**Comment by jiqing-feng** [comment]
Hi @echarlaix @fxmarty thanks for your comments. I have added `position_ids` on `ORTModelForCausalLM`. Would you please help to review it? Thanks.

**Comment by fxmarty** [comment]
Hi @jiqing-feng , could you comment a bit on your use case? I'm not sure what's the benefit of this in the ORT integration.

While `attention_mask` [is updated in `_update_model_kwargs_for_generation`](https://github.com/huggingface/transformers/blob/aab14120d4bbc64f0181c95434b64b92985ffd16/src/transformers/generation/utils.py#L761), `position_ids` is not. Thus, in my understanding, if we are to pass a `position_ids` to the `generate()` method, it would be kept as a constant at each `prepare_inputs_for_generation` call in the generation, e.g. [here](https://github.com/huggingface/transformers/blob/aab14120d4bbc64f0181c95434b64b92985ffd16/src/transformers/generation/utils.py#L1940). Which is something we don't wish.

Am I misunderstanding something @gante ?

**Comment by gante** [comment]
@fxmarty What I write next is from a `transformers` point of view :)

At the moment, `position_ids` is a required input to support left padding. For legacy reasons, we consider it an optional input the same way `attention_mask` is an optional input -- if not passed, we assume all tokens are valid, and `position_ids` is a `torch.arange`. In `.generate()`, we delegate its creation to the model's `prepare_inputs_for_generation`, where it is computed from the `attention_mask` (e.g. [here](https://github.com/huggingface/transformers/blob/3080fb714f11d28d8d3468548d4fa8a41e671a11/src/transformers/models/gpt2/modeling_gpt2.py#L1018)). 

This means that while `position_ids` is never passed into `.generate()`, is is passed from `.generate()` to the model and is updated at each iteration.

**Comment by fxmarty** [comment]
> In .generate(), we delegate its creation to the model's prepare_inputs_for_generation, where it is computed from the attention_mask (e.g. [here](https://github.com/huggingface/transformers/blob/3080fb714f11d28d8d3468548d4fa8a41e671a11/src/transformers/models/gpt2/modeling_gpt2.py#L1018)).

So this means that executing this controlflow is absolutely necessary, right?

**Comment by KexinFeng** [comment]
> Hi @jiqing-feng , could you comment a bit on your use case? I'm not sure what's the benefit of this in the ORT integration.
> 
> While `attention_mask` [is updated in `_update_model_kwargs_for_generation`](https://github.com/huggingface/transformers/blob/aab14120d4bbc64f0181c95434b64b92985ffd16/src/transformers/generation/utils.py#L761), `position_ids` is not. Thus, in my understanding, if we are to pass a `position_ids` to the `generate()` method, it would be kept as a constant at each `prepare_inputs_for_generation` call in the generation, e.g. [here](https://github.com/huggingface/transformers/blob/aab14120d4bbc64f0181c95434b64b92985ffd16/src/transformers/generation/utils.py#L1940) (since we'd go into [this piece of controlflow](https://github.com/huggingface/transformers/blob/9417c924af539be5f941c8a709a96b60dfe29eb3/src/transformers/models/gpt2/modeling_gpt2.py#L1024-L1025)). Which is something we don't wish.
> 
> Am I misunderstanding something @gante ?

@fxmarty Yeah, I was about to request for exactly the same feature. As @gante mentioned above, the `position_ids` (along with `attention_mask`) are necessary inputs to transformer based models like gpt2 to deal with left-padded input.  More specifically, my feature request comes from the context described in https://github.com/huggingface/optimum/issues/972. 


> > In .generate(), we delegate its creation to the model's prepare_inputs_for_generation, where it is computed from the attention_mask (e.g. [here](https://github.com/huggingface/transformers/blob/3080fb714f11d28d8d3468548d4fa8a41e671a11/src/transformers/models/gpt2/modeling_gpt2.py#L1018)).
> 
> So this means that executing this controlflow is absolutely necessary, right?

I don't think the controlflow of computing the `position_ids` from `attention_mask` is necessary, if this is what you referred to. As long as the traced transformer model (*.onnx) can take `position_ids` as an effective input, it should be good enough.

**Comment by fxmarty** [comment]
cc @echarlaix @michaelbenayoun

File: optimum/onnxruntime/modeling_decoder.py [hunk 1]
```diff
@@ -649,19 +650,22 @@ def forward(
                 input_ids=input_ids,
                 attention_mask=attention_mask,
                 past_key_values=past_key_values,
+                position_ids=position_ids,
```
Comment (michaelbenayoun) [inline optimum/onnxruntime/modeling_decoder.py:653]: `self.decoder` is an `ORTDecoder` right?
If so, we also need to update its forward method to handle `position_ids`.

File: optimum/onnxruntime/modeling_decoder.py [hunk 2]
```diff
@@ -649,19 +650,22 @@ def forward(
...
                 past_key_values=past_key_values,
                 attention_mask=attention_mask,
+                position_ids=position_ids,
             )
         else:
             outputs = self.decoder_with_past(
                 input_ids=input_ids[:, -1:],
                 past_key_values=past_key_values,
                 attention_mask=attention_mask,
+                position_ids=position_ids,
```
Comment (michaelbenayoun) [inline optimum/onnxruntime/modeling_decoder.py:668]: Same comment.

**Review by michaelbenayoun — COMMENTED** [review]
On my side, I am ok with adding this as long as we make sure we do not:

- Add any break changes
- Do not break `.generate`, it should work with and without the `position_ids` (except for the left padding case of course)

**Comment by jiqing-feng** [comment]
Hi @michaelbenayoun  , thanks for your comment. I have updated position_ids in ORTdecoder forward. Could you please review it? Thanks! cc @fxmarty @gante

**Comment by jiqing-feng** [comment]
> Hi everyone, @jiqing-feng @fxmarty @gante is there any updates on this feature?

cc @michaelbenayoun

**Comment by fxmarty** [comment]
@jiqing-feng @KexinFeng I now realize I misunderstood your argument and motivation for the PR. This appears to me to be a major bug in Optimum which should be fixed ASAP.
